src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,324 @@ using namespace opossum::expression_functional;  // NOLINT
+      for (const auto& predicate : join_node.join_predicates()) {
+        DebugAssert(predicate->type == ExpressionType::Predicate && predicate->arguments.size() == 2,
+                    "Expected binary predicate for join");
+        required_expressions.emplace(predicate->arguments[0]);
+        required_expressions.emplace(predicate->arguments[1]);
+      }
+    } break;
 
-  return consumed_columns;
+    // No pruning of the input columns to Delete, Update and Insert, they need them all.
+    case LQPNodeType::Delete:
+    case LQPNodeType::Insert:
+    case LQPNodeType::Update: {
+      const auto& left_input_expressions = node->left_input()->column_expressions();
+      required_expressions.insert(left_input_expressions.begin(), left_input_expressions.end());
+
+      if (node->right_input()) {
+        const auto& right_input_expressions = node->right_input()->column_expressions();
+        required_expressions.insert(right_input_expressions.begin(), right_input_expressions.end());
+      }
+    } break;
+  }
+
+  return required_expressions;
 }
 
-void ColumnPruningRule::_prune_columns_from_leaves(const std::shared_ptr<AbstractLQPNode>& lqp,
-                                                   const ExpressionUnorderedSet& referenced_columns) {
-  // Collect all parents of leaves and on which input side their leave is
-  // (if a node has two leaves as inputs, it will be collected twice)
-  auto leaf_parents = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, LQPInputSide>>{};
-  visit_lqp(lqp, [&](auto& node) {
-    // Early out for non-leaves
-    if (node->input_count() != 0) {
-      return LQPVisitation::VisitInputs;
-    }
+void recursive_gather_required_expressions(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t>& outputs_visited_by_node) {
+  const auto additional_columns = gather_required_expressions(node);
+  auto& required_expressions = required_expressions_by_node[node];
+  required_expressions.insert(additional_columns.begin(), additional_columns.end());
+
+  ++outputs_visited_by_node[node];
+  if (outputs_visited_by_node[node] < node->output_count()) return;
 
-    auto pruned_column_ids = std::vector<ColumnID>{};
-    for (const auto& expression : node->column_expressions()) {
-      if (referenced_columns.find(expression) != referenced_columns.end()) {
-        continue;
+  // Once all nodes that may require columns from this node (i.e., this node's outputs) have been visited, we can
+  // recurse into this node's inputs.
+  for (const auto& input : {node->left_input(), node->right_input()}) {
+    if (!input) continue;
+
+    // Make sure the entry exists, then insert all expressions that the current node needs
+    required_expressions_by_node[input];
+    for (const auto& required_expression : required_expressions) {
```

Comment:
Add the columns needed here (and above) if they come from the input node. Reasons why this might NOT be the case are: (1) The expression is calculated in this node (and is thus not available in the input node), or (2) we have two input nodes (i.e., a join) and the expressions comes from the other side.

Better?